### PR TITLE
JAMES-2418 Strong typing for repository URLs and keys

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    private final String value;
+    private final Protocol protocol;
+
+    private MailRepositoryUrl(String value) {
+        Preconditions.checkNotNull(value);
+        Preconditions.checkArgument(value.contains(":"), "':' is mandatory to delimit protocol");
+        this.value = value;
+        this.protocol = new Protocol(value.substring(0, value.indexOf(':')));
```
Comment (aduprat) [inline server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java]: value.split(":")[0]

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/Protocol.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.Objects;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+
+public class Protocol {
+    private final String value;
+
+    public Protocol(String value) {
+        Preconditions.checkNotNull(value);
```
Comment (aduprat) [inline server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/Protocol.java]: Strings.isNullOrEmpty
should throw when the protocol is empty, no?

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java (hunk 1 above)
Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java]: Or Splitter.on(":").splitToList(0)

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/Protocol.java (hunk 2 above)
Comment (chibenwa) [reply server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/Protocol.java]: As explained previously to Duc, an empty protocol, even if it makes littles sens is auhorized.  I don't want to introduce a regression here...

File: server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java (hunk 1 above)
Comment (rouazana) [reply server/mailrepository/mailrepository-api/src/main/java/org/apache/james/mailrepository/api/MailRepositoryUrl.java]: you mean Splitter.on(":") :)

**Comment by mbaechler** [comment]
merged

**Pull request closed without merging**